combinator/src/main/java/com/iluwatar/combinator/StudentValidation.java

```diff
@@ -0,0 +1,125 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.iluwatar.combinator;
+
+import java.text.DateFormat;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+final class ValidationSupport {
+    private static final ValidationResult valid = new ValidationResult(){
+        public boolean isValid(){ return true; }
+        public Optional<String> getReason(){ return Optional.empty(); }
+    };
+
+    static ValidationResult valid(){
+        return valid;
+    }
+}
+
+interface ValidationResult{
+    static ValidationResult valid(){
+        return ValidationSupport.valid();
+    }
+
+    static ValidationResult invalid(String reason){
+        return new Invalid(reason);
+    }
+
+    boolean isValid();
+
+    Optional<String> getReason();
+}
+
+final class Invalid implements ValidationResult {
+
+    private final String reason;
+
+    Invalid(String reason){
+        this.reason = reason;
+    }
+
+    public boolean isValid(){
+        return false;
+    }
+
+    public Optional<String> getReason(){
+        return Optional.of(reason);
+    }
+}
+
+public interface StudentValidation extends Function<Student, ValidationResult> {
+	static StudentValidation nameIsNotEmpty() {
+        return holds(student -> !student.name.trim().isEmpty(), "Name is empty.");
+    }
+
+    static StudentValidation surnameIsNotEmpty() {
+    	return holds(student -> !student.surname.trim().isEmpty(), "Surname is empty.");
+    }
+
+    static StudentValidation dateIsValid() {
+    	return holds(student -> CheckDate.isValid(student.birthDate), "Invalid birth date.");
+    }
+
+    static StudentValidation eMailContainsAtSign() {
+        return holds(student -> student.email.contains("@"), "Invalid e-mail. Missing @-sign.");
+    }
+
+    static StudentValidation isProgrammValid() {
+    	return holds(student -> Arrays.stream(Programm.values()).anyMatch((t) -> t.name().equals(student.stProgramm)), "Invalid student programm.");
+    }
+
+    static StudentValidation holds(Predicate<Student> p, String message){
+        return student -> p.test(student) ? ValidationResult.valid() : ValidationResult.invalid(message);
+    }
+
```

Comment:
Discussion: In current state, if we have another bean say, `Teacher` and we want to create `TeacherValidation` then will have to repeat code for `and` combinator.

I was wondering if we can move it to `ValidationSupport` and make `StudentValidation` extend `ValidationSupport`. We will have to generify `ValidationSupport`. Let me know you thoughts on it.